[SPARK-36073][SQL] EquivalentExpressions fixes and improvements

peter-toth opened the pull request:
### What changes were proposed in this pull request?

This PR:
- Fixes the performance issue mentioned in https://github.com/apache/spark/pull/32559/files#r633488455 and partially fixed in https://github.com/apache/spark/pull/33142/files#r660897248 using a new option to remove expressions from equivalence maps.
- Fixes a bug with identifying common expressions in conditional expressions (a side effect of the above new approach). After this PR, `add` will be common subexpression in the following example:
  ```
  val add = Add(Literal(1), Literal(2))
  val ifExpr1 = If(Literal(true), add, Literal(3))
  val ifExpr3 = If(GreaterThan(add, Literal(4)), Add(ifExpr1, add), Multiply(ifExpr1, add))
  var equivalence = new EquivalentExpressions
  equivalence.addExprTree(ifExpr3)
  ```
- Fixes a bug of transparently canonicalized expressions (like `PromotePrecision`) are considered common subexpressions.
  After this PR, `transparent` will not be common subexpression in the following example:
  ```
  val add = Add(Literal(1), Literal(2))
  val transparent = PromotePrecision(add)
  var equivalence = new EquivalentExpressions
  equivalence.addExprTree(transparent)
  ```

### Why are the changes needed?
Bugfix + performance improvement.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing + new UTs.

peter-toth committed:
[SPARK-36073][SQL] SubExpr elimination should include common child exprs of conditional expressions

SparkQA commented:
**[Test build #140852 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140852/testReport)** for PR 33281 at commit [`102942c`](https://github.com/apache/spark/commit/102942c49040c65ba8d009184b3fc818f8372d6e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45363/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45363/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45363/

peter-toth committed:
fix transparently canonicalized expressions

SparkQA commented:
**[Test build #140857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140857/testReport)** for PR 33281 at commit [`7786c0c`](https://github.com/apache/spark/commit/7786c0ccb8d7b716955de0a1e59acc822f0cb00c).

SparkQA commented:
**[Test build #140852 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140852/testReport)** for PR 33281 at commit [`102942c`](https://github.com/apache/spark/commit/102942c49040c65ba8d009184b3fc818f8372d6e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ExpressionStats(expr: Expression)(var useCount: Int) `

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45368/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140852/

peter-toth commented:
cc @cloud-fan, @Kimahriman, @maropu, @viirya

Kimahriman commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:
```diff
@@ -323,6 +323,34 @@ class SubexpressionEliminationSuite extends SparkFunSuite with ExpressionEvalHel
+    val ifExpr3 = If(GreaterThan(add, Literal(4)), Add(ifExpr1, add), Multiply(ifExpr1, add))
+
+    val equivalence = new EquivalentExpressions
+    equivalence.addExprTree(ifExpr3)
```
What was the behavior of this?

peter-toth replied:
This test actually proves that the new `updateExprTree(stats.expr, localEquivalenceMap, -stats.useCount)` approach fixes a bug as well.
Before this PR `equivalence.getAllExprStates(1)` didn't return anything because the `notChild` filter at https://github.com/apache/spark/pull/33281/files#diff-4d8c210a38fc808fef3e5c966b438591f225daa3c9fd69359446b94c351aa11eL90-L92 filtered out all child expressions (including `add`) of the common expression `ifExpr1`. But it should filter out only its children "defined by" `childrenToRecurse()` and `commonChildrenToRecurse()`. This new approach in the PR, that removes `ifExpr1` from `localEquivalenceMap`, keeps `add` in the map and so `add` is will have useCount = 2 in the end.

Kimahriman commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala:
```diff
@@ -323,6 +323,34 @@ class SubexpressionEliminationSuite extends SparkFunSuite with ExpressionEvalHel
+    val transparent = PromotePrecision(add)
+
+    val equivalence = new EquivalentExpressions
+    equivalence.addExprTree(transparent)
```
Does transparently canonicalized mean `PromotePrecision(add).canonicalized == add.canonicalized`?

peter-toth replied:
Yes, maybe I could rephrase this if it doesn't make sense.

viirya commented:
>  Fixes the performance issue mentioned in https://github.com/apache/spark/pull/32559/files#r633488455 and partially fixed in https://github.com/apache/spark/pull/33142/files#r660897248 using a new option to remove expressions from equivalence maps.
    Fixes an issue with identifying common expressions in conditional expressions (a side effect of the above).
    Fixes the issue of transparently canonicalized expressions (like PromotePrecision) are considered common subexpressions.

Could you explain more about the three points listed in the description?

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -70,34 +96,39 @@ class EquivalentExpressions {
+    // its children recursively if required) from `localEquivalenceMap`. The remaining highest
+    // expression in `localEquivalenceMap` is also common expression.
+    var statsOption = Some(localEquivalenceMap).filter(_.nonEmpty).map(_.maxBy(_._1.height)._2)
+    while (statsOption.nonEmpty) {
+      val stats = statsOption.get
+      updateExprTree(stats.expr, localEquivalenceMap, -stats.useCount)
```
Oh, I got your idea here. For performance, is there significant difference? The current approach is simpler. The filtering in `localEquivalenceMap` is based on height so it should be very fast. Is this still a performance bottleneck?

viirya replied:
My concern is that this change complicates the computation of `useCount`. It will be harder to debug in the future. Before we are not certain that this is a real performance bottleneck, this may look like a premature optimization.

peter-toth replied:
Honestly I'm not sure how significant the difference is. I think if we have deep expressions in `localEquivalenceMap` then filtering by height (before this PR) might not help a lot.
The new code in this PR might look a bit complex at first, but actually it is very simple, we just remove expressions from the `localEquivalenceMap` with the reverse of `addExprTree()`.

This new approach also fixes a bug tested here: https://github.com/apache/spark/pull/33281/files#r667343014

viirya replied:
I see. I'd rather consider it as an improvement as it doesn't cause query failure or codegen failure, though it fails to identify a common subexpression.

That's said, we don't need to hurry on this for 3.2.

peter-toth replied:
All right, I've changed ticket type to improvement.

peter-toth commented:
> > Fixes the performance issue mentioned in https://github.com/apache/spark/pull/32559/files#r633488455 and partially fixed in https://github.com/apache/spark/pull/33142/files#r660897248 using a new option to remove expressions from equivalence maps.
> > Fixes an issue with identifying common expressions in conditional expressions (a side effect of the above).
> > Fixes the issue of transparently canonicalized expressions (like PromotePrecision) are considered common subexpressions.
> 
> Could you explain more about the three points listed in the description?

Sure. I've updated the description and added some examples.
If we consider this PR too complex then maybe I should split the 2 commits into 2 separate PRs. The first commit belongs to the first 2 points, the second commit belongs to the 3rd.

peter-toth commented:
Gentle ping @viirya, @cloud-fan.

peter-toth commented:
@viirya, @cloud-fan I wonder if we can move forward with this?

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -38,19 +40,43 @@ class EquivalentExpressions {
+            false
+          } else {
+            // Should not happen
+            throw new IllegalArgumentException(
```
Let's follow SPARK-33539 and add a new method in `QueryExecutionErrors`.

peter-toth replied:
Thanks @allisonwang-db. Added it in https://github.com/apache/spark/pull/33281/commits/227cad1af875d9c432fae9a7428d23d42cb4392e

peter-toth committed:
fix comments and review findings

SparkQA commented:
**[Test build #145034 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145034/testReport)** for PR 33281 at commit [`227cad1`](https://github.com/apache/spark/commit/227cad1af875d9c432fae9a7428d23d42cb4392e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49506/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49506/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49506/

SparkQA commented:
**[Test build #145034 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145034/testReport)** for PR 33281 at commit [`227cad1`](https://github.com/apache/spark/commit/227cad1af875d9c432fae9a7428d23d42cb4392e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145034/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -38,19 +41,41 @@ class EquivalentExpressions {
-  private def addExprToMap(
-      expr: Expression, map: mutable.HashMap[ExpressionEquals, ExpressionStats]): Boolean = {
+  /**
+   * Adds or removes an expression to/from the map and updates `useCount`.
```
how to indicate if we need to do add or remove?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -38,19 +41,41 @@ class EquivalentExpressions {
+            map -= wrapper
+            false
+          } else {
+            // Should not happen
```
If this should not happen, I think throwing `IllegalStateException` is better as it's a bug. `QueryExecutionErrors` is for user-facing errors.

peter-toth replied:
Throwing `IllegalStateException` looks reasonable to me. Fixed in https://github.com/apache/spark/pull/33281/commits/7c4d5065d7af78a6cd77788a6c2b3a9dcb9f88bf

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -38,19 +41,41 @@ class EquivalentExpressions {
+            map.put(wrapper, ExpressionStats(expr)(useCount))
+          } else {
+            // Should not happen
+            throw QueryExecutionErrors.updateEquivalentExpressionsError(expr, map, useCount)
```
ditto

peter-toth replied:
Fixed in https://github.com/apache/spark/pull/33281/commits/7c4d5065d7af78a6cd77788a6c2b3a9dcb9f88bf

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:
```diff
@@ -59,45 +84,43 @@ class EquivalentExpressions {
-   * Adds only expressions which are common in each of given expressions, in a recursive way.
-   * For example, given two expressions `(a + (b + (c + 1)))` and `(d + (e + (c + 1)))`,
-   * the common expression `(c + 1)` will be added into `equivalenceMap`.
+   * Adds or removes only expressions which are common in each of given expressions, in a recursive
```
It may remove expressions from the given `map`? I think it only adds.

peter-toth committed:
fix exception

SparkQA commented:
**[Test build #145069 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145069/testReport)** for PR 33281 at commit [`c7c7016`](https://github.com/apache/spark/commit/c7c7016147abbf4c88c0b22b626ea6fd6b927413).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49538/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49538/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49538/

SparkQA commented:
**[Test build #145069 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145069/testReport)** for PR 33281 at commit [`c7c7016`](https://github.com/apache/spark/commit/c7c7016147abbf4c88c0b22b626ea6fd6b927413).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145069/

cloud-fan commented:
thanks, merging to master!

peter-toth commented:
Thanks for the review @cloud-fan!
